geode-core/src/test/java/org/apache/geode/internal/cache/map/RegionMapPutTest.java

```diff
@@ -89,13 +106,307 @@ private RegionEntry doPut() {
+    givenExistingRegionEntry();
+    givenAnOperationThatDoesNotGuaranteeOldValue();
+    givenPutNeedsToDoCacheWrite();
+
+    Object oldValue = new Object();
+    when(existingRegionEntry.getValueRetain(same(internalRegion), eq(true))).thenReturn(oldValue);
+
+    doPut();
+
+    verify(event, times(1)).setOldValue(same(oldValue));
+    verify(event, never()).setOldValue(not(same(oldValue)));
+  }
+
+  @Test
+  public void setsEventOldValueToNotAvailable_ifIsCacheWriteAndOperationDoesNotGuaranteeOldValue_andExistingValueIsNull() {
+    givenPutNeedsToDoCacheWrite();
+    givenAnOperationThatDoesNotGuaranteeOldValue();
+    givenExistingRegionEntry();
+
+    when(existingRegionEntry.getValueRetain(same(internalRegion), eq(true))).thenReturn(null);
+
+    doPut();
+
+    verify(event, times(1)).setOldValue(Token.NOT_AVAILABLE);
+  }
+
   @Test
   public void retrieveOldValueForDeltaDefaultToFalse() {
     createInstance();
 
     assertThat(instance.isRetrieveOldValueForDelta()).isFalse();
   }
 
+
+  @Test
+  public void eventOldValueNotAvailableCalled_ifCacheWriteNotNeededAndNotInitialized() {
+    givenPutDoesNotNeedToDoCacheWrite();
+    when(internalRegion.isInitialized()).thenReturn(false);
+
+    doPut();
+
+    verify(event, times(1)).oldValueNotAvailable();
+  }
+
+  @Test
+  public void eventOperationNotSet_ifCacheWriteNeededAndInitializedAndReplaceOnClient() {
+    givenPutNeedsToDoCacheWrite();
+    when(internalRegion.isInitialized()).thenReturn(true);
+    givenReplaceOnClient();
+
+    doPut();
+
+    verify(event, never()).makeCreate();
+    verify(event, never()).makeUpdate();
+  }
+
+  @Test
+  public void putExistingEntryCalled_ifReplaceOnClient() throws RegionClearedException {
+    givenPutDoesNotNeedToDoCacheWrite();
+    when(internalRegion.isInitialized()).thenReturn(true);
+    givenReplaceOnClient();
+
+    doPut();
+
+    verify(event, times(1)).putExistingEntry(any(), any(), anyBoolean(), any());
+  }
+
+  @Test
+  public void eventOperationMadeCreate_ifCacheWriteNeededAndInitializedAndNotReplaceOnClientAndEntryRemoved() {
+    givenPutNeedsToDoCacheWrite();
+    when(internalRegion.isInitialized()).thenReturn(true);
+    givenReplaceOnPeer();
+    when(createdRegionEntry.isDestroyedOrRemoved()).thenReturn(true);
+
+    doPut();
+
+    verify(event, times(1)).makeCreate();
+    verify(event, never()).makeUpdate();
+  }
+
+  @Test
+  public void eventOperationMadeUpdate_ifCacheWriteNeededAndInitializedAndNotReplaceOnClientAndEntryExists() {
+    givenPutNeedsToDoCacheWrite();
+    when(internalRegion.isInitialized()).thenReturn(true);
+    givenReplaceOnPeer();
+    when(createdRegionEntry.isDestroyedOrRemoved()).thenReturn(false);
+
+    doPut();
+
+    verify(event, never()).makeCreate();
+    verify(event, times(1)).makeUpdate();
+  }
+
+  @Test
+  public void cacheWriterBeforePutNotCalled_ifNotInitialized() {
```

Comment:
Should this (and in the tests below) be `cacheWrite…` instead of `cacheWriter…`?

Reply:
Yes!